### DISCUSSION TITLE
Add `filter_by_criteria` to `QueryFilter`?

### DISCUSSION THREAD

**Discussion by JonasCir** [SeaQL/sea-orm#1360] [Ideas]
### User story

As a dev using `sea-orm`, I want to create query filters for entities (e.g., based on `GET` parameters) conveniently.

### Status quo

Let's stick to the example of `GET` parameters (which is also motivated by a question on Discord).  The status quo is that in the request handler, the parameters needs to be unpacked and then are either:
1. are passed individually to a backend function as optional arguments where a long `if Some(x)` cascade will craft the query. This can result in quite large method signatures which needs to be changed every time a new parameter is added. This needs to be repeated for every entity type.
2. are copied in a dedicated struct (e.g., `PersonQuery`) which records what fields are queried in which form. While the method signature for the backend call is clean, this approach again requires a lot of unpacking, matching and unnecessary copying between structs. This needs to be repeated for every entity type, although traits and added derives might help a bit.
3. are directly "translated" to a query with a `if Some(x)` or `match` cascade , which bloats up the handler code significantly as seen in the Discord example.

### Proposal

I propose a `Criteria` approach similar to other ORMs, where more complex filters can be easily created for an entity and devs do not need to roll their own glue code.

1. Introduce trait `CriteriaFilter`:
 ```rust
pub trait CriteriaFilter : IntoCondition;
 ```
2. Adjust `QueryFilter`:
 ```rust
fn filter_by_criteria<C>(self, criteria: C) -> Self where C: CriteriaFilter;
 ```
3. Introduce new macro logic (i.e., `DeriveCriteria`) which spawns a new struct implementing `CriteriaFilter` for a given `Model`:
 ```rust
#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, DeriveCriteria)]
#[sea_orm(table_name = "person")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub name: String,
    pub age: i32,
}

// Derive yields something like
#[derive(Default))
pub struct PersonCriteria {
  pub condition_type: ConditionType // ANY or ALL Enum
  pub id: Option<I32Criteria>,
  pub name: Option<StringCriteria>,
  pub age: Option<I32Criteria>,
}

impl CriteriaFilter for PersonCriteria {
  fn into_condition(self) -> Condition {
    let cond = condition_type.new_cond(); // returns either Condition created with all() or any()

    self.id.map(|id| cond.add(id.condition_for_col(Column::Id)));
    self.name.map(|name| cond.add(name.condition_for_col(Column::Name));
    self.age.map(|age| cond.add(age.condition_for_col(Column::Age)));
    
    cond
  }
}

// in our GET handler pseudo code

pub async fn get_persons(
    State(state): State<AppState>,
    params: Query<GetAllPersons>,
) -> Json<Vec<entity::person::Person>> {

    let criteria = PersonCriteria {
        condition_type: ConditionType::All,
        id: params.get("id").map(|id| I32Criteria::Gt(id))), // greater than Id
        name: params.get("name").map(|name| StringCriteria::Like(name))), // SQL LIKE string search 
        .. Default::default() // we do not care about age and set it per default to None
    }

     entity::persons::Entity::find_by_criteria(criteria)
                    .limit(params.limit.unwrap_or(42))
                    .all(&state.database)
                    .await
                    .expect("Failed to get persons")
    // or pass the criteria to a backend method for further access checks or whatever 
}
 ```
4. this needs to following in addition:
 ```rust
pub enum ConditionType {
  All,
  Any,
}

impl ConditionType {
  fn new_cond(self) -> Condition {
    match self {
      All => Cond::all(),
      Any => Cond::any(),  
  }
}

pub trait CriteriaCondition {
  fn condition_for_col(self, col: $don't know the type sorry$) -> Condition;
}

 pub enum I32Criteria {
   Gt(i32) // greater than
   Le(i32) // etc   
 }
 
impl CriteriaCondition for I32Criteria {
  fn condition_for_col(self, column: some_type) -> Condition {
   match self {
      Gt(val) => column.gt(val),
      Le(val) => column.le(val),
   }  
  }
}

 // Same for StringCriteria with Like, ILike etc enum variants...
```

### Open Questions
* How to deal with negations? I thought of adding an additional Enum like `IsNot` which wraps the criteria fields in `Is` or `Not` in a model criteria. Construction would be `Is(I32Criteria::Eq(42))` or `Not(StringCriteria::Like("some_string")))`. `CriteriaCondition` could be simply implemented for `IsNot` which calls the inner `condition_for_col` and adds a `.not()` in case of negation. I think this could be better than having a `I32Criteria::NotGt` and the like just to cover negation.
* Do we want to expose the fields as pub or do we want a builder style pattern?
* We could just implement `Into<Condition>` and use `filter` directly. Do we even need to introduce `CriteriaFilter` trait and `filter_by_criteria`? Not certain what is easier to understand/teach to users.

**Comment by billy1624** [comment]
Hey @JonasCir, thanks for the write up. I think what you mention is actually similar to the filter in seaography.

https://github.com/SeaQL/seaography/blob/d64421783e6d3b2656c119ca0cbedbd53d5f2e22/examples/mysql/src/entities/actor.rs#L38-L193

We can provide a derive macro on the Model struct to implement the filter struct and related traits.

https://github.com/SeaQL/seaography/blob/d64421783e6d3b2656c119ca0cbedbd53d5f2e22/derive/src/filter.rs

```rs
pub struct Filter {
    pub or: Option<Vec<Box<Filter>>>,
    pub and: Option<Vec<Box<Filter>>>,
    id: Option<<i32 as FilterTrait>,
    name: Option<<String as FilterTrait>,
    age: Option<<i32 as FilterTrait>,
}
```

**Reply by JonasCir** [reply]
Hi @billy1624, that's great, thank you for the hint! If you don't mind, I would go ahead and create an issue to work on it? Implementation would then (mostly) mirror what you just showed here.

**Reply by billy1624** [reply]
Please proceed!
Let me know if you need help along the way :)

**Comment by 112RG** [comment]
+1 for this. Would really help clean up my match hell https://github.com/Ortygia/Deaftone/blob/5c8fced07b8885f8301503786f5ac246fa89a756/src/handlers/albums.rs#L104

**Comment by lazulit3** [comment]
If it helps anyone, I'll share an example of a naive/simplistic approach I used for turning query string (`GET` filter) parameters into a `Condition` for filtering on an Entity. In the first pass it only matches exact values & uses `Condition::all()` (without support for `any`) and needs a lot of work.

I appreciate that this proposal captures the more complex/powerful side of filtering (e.g. any vs all, different operators for filtering), as I'll need to tackle that down the road. :)

<details>
<summary>Expand me for naive / simple example of filtering (turning query strings into Conditions)</summary>
Previously I had this kind of manual approach:

```rust
pub async fn get_drugs(
    Query(params): Query<HashMap<String, String>>,
    State(db): State<DatabaseConnection>,
) -> Result<Json<Vec<drug::Model>>, ApiError> {
    let condition = params
        .get("name")
        .map(|name| Condition::all().add(drug::Column::Name.eq(name)));

    let drugs = select_drugs_with_condition(&db, condition)
        .await
        .context("Failed to get drugs")?;
    Ok(Json(drugs))
```

I recently [replaced](https://github.com/lazulit3/edihkal/pull/47/commits/bbd51aa3e92337b725ebe895891f5c5a2e898382) this with usage that looks like this:

`QueryParams` is implemented for some `EntityTrait` to define how the query parameter (key) maps to an entity's `Column` variants. (This also enforces some validation around which columns can be filtered on.) Although this mapping is no less manual than the past iteration, it's at least defined once per entity and not directly where used in handler(s):

```rust
pub trait QueryParams: EntityTrait {
    fn column(query_param: &str) -> Option<Self::Column>;
}

impl QueryParams for Drug {
    fn column(query_param: &str) -> Option<Self::Column> {
        match query_param {
            "name" => Some(drug::Column::Name),
            _ => None,
        }
    }
}
```

Then a `Filters` type (with some Entity type generic) is constructed from the query string (and this `Filters` implements `IntoCondition`):

```rust
pub async fn get_drugs(
    Query(query): Query<HashMap<String, String>>,
    State(db): State<DatabaseConnection>,
) -> Result<Json<Vec<drug::Model>>, ApiError> {
    let filter = Filters::<Drug>::from_query_string(query);

    let drugs = select_drugs(&db, filter).await.context("Failed to get drugs")?;
    
    Ok(Json(drugs))
}
```

`Filters<E>::from_query_string(...)` uses `QueryParams` to map the query params/values to the `Entity::Column` variants & values, and implements `IntoCondition` to fold those into a `Condition::all()`:

```rust
pub struct Filters<E: QueryParams + EntityTrait>(Vec<(E::Column, String)>);

impl<E: QueryParams + EntityTrait> Filters<E> {
    pub fn from_query_string(query_string: HashMap<String, String>) -> Self {
        Self(
            query_string
                .into_iter()
                .filter_map(|(param, value)| E::column(&param).map(|column| (column, value)))
                .collect(),
        )
    }
}

impl<E: QueryParams> IntoCondition for Filters<E> {
    fn into_condition(self) -> Condition {
        self.0.into_iter().fold(Condition::all(), |all, (column, value)| {
            all.add(column.eq(value))
        })
    }
}
```
</details>

**Reply by tyt2y3** [reply]
I like your approach, especially the opt-in part of it. Anything more complicated will evolve into something like the Lucene Query syntax. Or we will end up in GraphQL, which incidentally, is what we are trying to do in Seaography.
